### PR TITLE
modify _TopKGrad so that all operations can run on GPU for better performance

### PR DESCRIPTION
**Pull request by carusyte**
This PR is intended to be a workaround to address #5719 by working on the Python client code level.
Recap:
1. Upcasting tensor to int64 type before sending to `array_ops.gather`, due to #5445;
2. Replace `sparse_ops.sparse_to_dense` with `array_ops.scatter_nd`, due to #8321;

### CODE REVIEW

File: .gitignore
Comment (ebrevdo) [inline .gitignore]: Remove these changes

File: tensorflow/python/ops/nn_grad.py
Comment (ebrevdo) [inline tensorflow/python/ops/nn_grad.py]: Remove commented out code.

**Review by ebrevdo — CHANGES_REQUESTED** [review]
Looks great

**Comment by carusyte** [comment]
The modification has been done.

**Comment by carusyte** [comment]
What should I do now? Is there any further modifications that need to be made? This is the first time I submit PR to tensorflow, guidance may be needed...

File: tensorflow/python/ops/nn_grad.py
Comment (ebrevdo) [inline tensorflow/python/ops/nn_grad.py]: Here can you just use [...] Instead of reshape(..., [1])?

**Review by ebrevdo — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/nn_grad.py
Comment (carusyte) [reply tensorflow/python/ops/nn_grad.py]: Sure! Didn't realize I could "shape" a rank 1 tensor in such a neat way!

**Review by ebrevdo — DISMISSED** [review]

**Comment by ebrevdo** [comment]
@tensorflow-jenkins  test this please.

**Comment by carusyte** [comment]
I just got upgraded to fathership ver 1.0.0 today and was busy looking after my wife and newborn when I missed a lot that had happened in this PR... Thanks for everyone who's moving this forward.

Now during my break I noticed that the PR got stuck by a `Ubuntu contrib - Internal CI build`, which I don't quite understand... I looked into the details and found the `mangled stack trace`, which 
 perplexed me further. I can't find any clue related to the `_TopKGrad` modification, but I got this relatively "readable" stack trace:

    *** Begin stack trace ***
	tensorflow::CurrentStackTrace()

	gsignal
	abort

	__libc_malloc
	opendir
	tensorflow::PosixFileSystem::GetChildren(std::string const&, std::vector<std::string, std::allocator<std::string> >*)
	tensorflow::internal::GetMatchingPaths(tensorflow::FileSystem*, tensorflow::Env*, std::string const&, std::vector<std::string, std::allocator<std::string> >*)
	tensorflow::PosixFileSystem::GetMatchingPaths(std::string const&, std::vector<std::string, std::allocator<std::string> >*)
	tensorflow::Env::GetMatchingPaths(std::string const&, std::vector<std::string, std::allocator<std::string> >*)
	tensorflow::RestoreV2::Compute(tensorflow::OpKernelContext*)
	tensorflow::ThreadPoolDevice::Compute(tensorflow::OpKernel*, tensorflow::OpKernelContext*)
	
    Eigen::NonBlockingThreadPoolTempl<tensorflow::thread::EigenEnvironment>::WorkerLoop(int)
	std::_Function_handler<void (), tensorflow::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}>::_M_invoke(std::_Any_data const&)

	clone
    *** End stack trace ***

I'm not sure if this is a common issue lurking in the master code right now, or how this stack trace can be linked to the `_TopKGrad`, will get back to it and try to find out when I time permitted.

**Comment by ebrevdo** [comment]
Doesn't look related. Was it during running a topk test?

**Comment by carusyte** [comment]
the test target points to `//tensorflow/contrib/data/python/kernel_tests/serialization:optimize_dataset_serialization_test`

full error log: [link](https://source.cloud.google.com/results/invocations/80e91e52-4e14-46ca-bef4-aac835bf4508/targets/%2F%2Ftensorflow%2Fcontrib%2Fdata%2Fpython%2Fkernel_tests%2Fserialization:optimize_dataset_serialization_test/log)




>Doesn't look related. Was it during running a topk test?

**Comment by ebrevdo** [comment]
That's unrelated to you, feel free to ignore

**Comment by carusyte** [comment]
Emmmm... I've got some time these days and I'd like to move on this PR. Since the failed test is not related to this particular PR, is it able to skip the test and carry out the other required procedure? @ebrevdo

**Comment by carusyte** [comment]
@ebrevdo @drpngx Sorry to bother again. Any further guidance/instructions/recommendations on how to move this PR forward? Or Is it common for PRs for issues labeled as "feature request" to be put at low priority? Or is there better solutions under way?

**Comment by drpngx** [comment]
Not sure, @ebrevdo good to go?

**Review by ebrevdo — DISMISSED** [review]

**Comment by carusyte** [comment]
@ebrevdo I found an error report in _Ubuntu Sanity_ check indicating an unused import of `sparse_ops ` in the `nn_grad.py`. I made a new commit to resolve this issue, but I haven't squashed all these commits in order to preserve @drpngx 's commit, if it is advised to do it this way.

The former code review has been dismissed as stale automatically, could you please make another review and initiate the checks one more time? Sorry for all the inconvenience.

**Review by ebrevdo — APPROVED** [review]

**Comment by carusyte** [comment]
@ebrevdo Thank you for taking the time to review this PR! It's very encouraging to see that it has passed all the checks except one - the _import/copybara_ check, which is also tagged as **Required** as I've noticed... Seems that this check report has been hanging there for more than 12 hours. I'm wondering if this is a normal state, but I don't find any entry to inspect the progress of it. So far, all that I can do is wait?

**Comment by drpngx** [comment]
Thanks @carusyte for checking, we were waiting for the CI tests to finish before we initiate the merge. Now it's `ready to pull` and should be pushed, as long as the internal tests pass. Stay tuned.

**Comment by carusyte** [comment]
@drpngx Thanks for the clarification of the process. Now I'm less confused. 
One more question, I've been trying to track the progress of __import/copybara__ check (is it the same thing as the internal test you've mentioned?), by following the [Details](http://go/import-review/9781) link, but it directs me to an invalid url address, maybe it's an internal url or something. Anyway, is it a manual or automated process? How long would it take, usually speaking?

**Comment by drpngx** [comment]
I think you should be able to see the status `Import/copybara` as it transitions from one state to the other. It's a manual, multi-step process of reviewing the changes for security, running tests, and getting approvals. It's maybe one hour in the fastest case. In this instance, I forgot to check back and approve the first level. After that, it will need to pass tests and get another approval.

**Comment by carusyte** [comment]
Ops...the `import/copybara` reported `An error happened while migrating the change`, although this PR seemed to have been merged. I can't find any entry to inspect this error. Is it serious?

**Comment by drpngx** [comment]
I think it's probably fine. Please check that [the merged commit](https://github.com/tensorflow/tensorflow/commit/229652512f745b9ef110c02deb9dce98fd33b5be) has all your changes.

**Pull request merged**
